### PR TITLE
XGboost genotype classifcation random search

### PR DESCRIPTION
**Pull request by MattsonCam**
In this pr, xgboost is optimized using random search. Please let me know if you have any suggestions.

### CODE REVIEW

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+data_path.mkdir(parents=True, exist_ok=True)
+
+
+# # Define Functions
+
+# In[5]:
+
+
+def filter_features(X_data):
+    """
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:78]: Might be worth adding in the docstring here a brief rationale for why filtering is important - you can also consider adding the `like` elements below as a function argument and looping through the inplace drop.

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        Cell data to be down-sampled by genotype.
+
+    Returns
+    -------
+    pandas Dataframe
+        Cell data down-sampled by genotype.
+    """
+
+    min_gene = df[gene_column].value_counts().min()
+    return (df.groupby(gene_column, group_keys=False)
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:114]: consider exposing the random state variable as a function argument to enable a future user to generate different random samples by toggling random state

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    platedf.groupby(["Metadata_genotype", "Metadata_Well"])
+    .size().reset_index(name="Metadata_cell_count")
+)
+
+
+# ## Split Data
+
+# ### Seperate the test (holdout) cell data
+# The number of cells in the train-validation set is maximized by minimizing the number of cells sampled from the adjusted minority class for a desired number of holdout wells.
+# This also adjusts the number of cells in the holdout wells for the classes other than the minority class to minimize the number of cells that would otherwise not be included in any of the datasets.
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:147]:
```suggestion
# We aim to maximize the the number of cells in the train-validation set.
# We achieve this by select the specific holdout wells that maximize the minority class in the train-validation set.
# In other words, we choose the combination of wells for train-validation that, together, include the highest number of cells in the genotype category that has the fewest number of cells.
# By side-effect, this process also minimizes the number of cells dropped from training in our downsampling procedure to balance datasets for class size prior to model training.
```

I rephrased this to increase clarity - Please double check my logic here.

Specifically, I split sentences with complex thoughts to simplify the message. I also highlight the effect of what we did in addition to how we do it.

Comment (gwaybio) [reply 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:147]: (also, I just realize i made this suggestion in an nbconverted file - feel free to transfer final version back to ipynb prior to re-nbconverting)

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+# In[9]:
+
+
+kwargs = {
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:152]: consider increasing clarity of this dictionary by improving variable name. What kind of kwargs are they? Perhaps `model_variable_kwargs` is a more appropriate name?

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+# Load normalized qc data
+platedf_path = pathlib.Path(root_dir / "nf1_painting_repo/3.processing_features/data/single_cell_profiles/Plate_5_sc_normalized_qc.parquet").resolve(strict=True)
+
+platedf = pd.read_parquet(platedf_path)
+
+# Access spliting utilities
+sys.path.append(f"{root_dir}/1.train_models/utils")
+feat_cols = load("selected_features_1_qc_prefs.joblib")
+from GreedySubsetSum import GreedySubsetSum
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:58]: Consider moving this import (and corresponding sys path command) to the import section of this module) to increase readability

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py (hunk 4 above)
Comment (gwaybio) [reply 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:152]: oh, maybe `greedy_subset_well_kwargs`?

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+
+
+# # Train Models
+
+# ## Define parameters
+
+# In[16]:
+
+
+param_ranges = {
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:220]: Consider adding rationale of why you chose this specific range

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+rsrdf["max_test_accuracy_mean"] = 1 - rsrdf["min_test_merror_mean"]
+
+
+# ## Save random search results
+
+# In[21]:
+
+
+rsrdf.to_csv(f"{data_path}/random_search_xgb_opt_results.tsv", sep="\t")
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:310]: love to see this output!

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+
+# ## Save split indices
```
Comment (gwaybio) [inline 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:313]: I'm a bit confused as to what this is? Why is doing this important? As is, it is also hard to determine if there are any explicit links to the split_indices file and the actual results - why do you need both? Is there a way to confirm alignment between indices and samples?

**Review by gwaybio — APPROVED** [review]
Looks great! Only minor comments aimed at increasing clarity - no need for me to review again after you decide what to do with my feedback.

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py (hunk 2 above)
Comment (MattsonCam) [reply 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:114]: I'll definetly consider this in the future

File: 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py (hunk 3 above)
Comment (MattsonCam) [reply 1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py:147]: I like the way you phrased this. I added this in other nf1 code

**Comment by MattsonCam** [comment]
Since creating this pr I've tried other approaches, which seem to be better in terms of performance and compute time. This model was useful to explore the capabilities of the XGBoost framework, but I think it is redundant in the code base at this time given the most recent model. Therefore, I will close this pr. The suggestions were also very helpful, which I will keep in mind for future prs.

**Pull request closed without merging**